ledger/src/blockstore_processor.rs

```diff
@@ -921,54 +924,97 @@ pub fn confirm_slot(
+            // Either way, check the result of the signature verification.
+            let transaction_check_result = check_result.finish_verify();
+
+            match verifier {
+                Some(mut verifier) => {
+                    let verified = verifier.finish_verify();
+                    timing.poh_verify_elapsed += verifier.poh_duration_us();
+                    // The GPU Entry verification (if any) is kicked off right when the CPU-side
+                    // Entry verification finishes, so these times should be disjoint
+                    timing.transaction_verify_elapsed +=
+                        transaction_cpu_duration_us + check_result.gpu_verify_duration();
+                    if !verified {
+                        warn!("Ledger proof of history failed at slot: {}", bank.slot());
+                        return Err(BlockError::InvalidEntryHash.into());
+                    } else if !transaction_check_result {
+                        warn!("Ledger proof of history failed at slot: {}", bank.slot());
+                        return Err(TransactionError::SignatureFailure.into());
+                    }
+                }
+                _ => {
+                    if !transaction_check_result {
+                        warn!("Ledger proof of history failed at slot: {}", bank.slot());
+                        return Err(TransactionError::SignatureFailure.into());
+                    }
+                }
+            }
 
-    let mut replay_elapsed = Measure::start("replay_elapsed");
-    let mut execute_timings = ExecuteTimings::default();
-    let cost_capacity_meter = Arc::new(RwLock::new(BlockCostCapacityMeter::default()));
-    // Note: This will shuffle entries' transactions in-place.
-    let process_result = process_entries_with_callback(
-        bank,
-        &mut entries,
-        true, // shuffle transactions.
-        entry_callback,
-        transaction_status_sender,
-        replay_vote_sender,
-        &mut execute_timings,
-        cost_capacity_meter,
-    )
-    .map_err(BlockstoreProcessorError::from);
-    replay_elapsed.stop();
-    timing.replay_elapsed += replay_elapsed.as_us();
+            process_result?;
 
-    timing.execute_timings.accumulate(&execute_timings);
+            progress.num_shreds += num_shreds;
+            progress.num_entries += num_entries;
+            progress.num_txs += num_txs;
+            if let Some(last_entry_hash) = last_entry_hash {
+                progress.last_entry = last_entry_hash;
+            }
 
-    if let Some(mut verifier) = verifier {
-        let verified = verifier.finish_verify();
-        timing.poh_verify_elapsed += verifier.poh_duration_us();
-        timing.transaction_verify_elapsed += transaction_duration_us;
-        if !verified {
+            Ok(())
+        }
+        Err(err) => {
+            if let Some(mut verifier) = verifier {
+                verifier.finish_verify();
```

Comment:
Sorry can you go back to bailing early? I was wrong earlier (https://github.com/solana-labs/solana/pull/20851#discussion_r756487494) and I think it's fine if the GPU thread finishes lazily, no need to block on that.